packages/core/contracts/zodiac/OptimisticOracleModule.sol

```diff
@@ -0,0 +1,232 @@
+        // create a proposal with a bundle of transactions
+        // note: based in part on the UMA Governor contract
+        // https://github.com/UMAprotocol/protocol/blob/master/packages/core/contracts/oracle/implementation/Governor.sol
+        // note: optional explanation explains the intent of the transactions to make comprehension easier
+        uint256 id = proposals.length;
+        uint256 time = block.timestamp;
+        address proposer = msg.sender;
+
+        // Add a zero-initialized element to the proposals array.
+        Proposal storage proposal = proposals.push();
+        proposal.requestTime = time;
+        proposal.ancillaryData = _explanation;
+
+        // Construct the ancillary data.
+        AncillaryData.appendKeyValueUint(proposal.ancillaryData, "id", id);
+        AncillaryData.appendKeyValueAddress(proposal.ancillaryData, "module", address(this));
+
+        // Initialize the transaction array.
+        for (uint256 i = 0; i < _transactions.length; i++) {
+            require(_transactions[i].to != address(0), "The `to` address cannot be 0x0");
+            // If the transaction has any data with it the recipient must be a contract, not an EOA.
+            if (_transactions[i].data.length > 0) {
+                require(_isContract(_transactions[i].to), "EOA can't accept tx with data");
+            }
+            proposal.transactions.push(_transactions[i]);
+        }
+
+        // Propose a set of transactions to the OO. If not disputed, they can be executed with executeProposal().
+        // docs: https://github.com/UMAprotocol/protocol/blob/master/packages/core/contracts/oracle/interfaces/SkinnyOptimisticOracleInterface.sol
+        _getOptimisticOracle().requestAndProposePriceFor(
+            identifier,
+            uint32(time),
+            proposal.ancillaryData,
+            collateral,
+            0,
+            bond,
+            uint256(liveness),
+            proposer,
+            // Canonical value representing "True"; i.e. the transactions are valid.
+            int256(1e18)
+        );
+
+        emit TransactionsProposed(id, proposer, time);
+    }
+
+    function executeProposal(uint256 _proposalId, uint256 _transactionIndex) public payable nonReentrant() {
+        // execute transactions in an approved proposal using exec() function
+        Proposal storage proposal = proposals[_proposalId];
+
+        // this will revert if the price has not settled
+        int256 price = _getOracle().getPrice(identifier, proposal.requestTime, proposal.ancillaryData);
+
+        Transaction memory transaction = proposal.transactions[_transactionIndex];
+
+        require(
+            _transactionIndex == 0 || proposal.transactions[_transactionIndex - 1].to == address(0),
+            "Previous tx not yet executed"
+        );
+        require(transaction.to != address(0), "Tx already executed");
+        require(price != 0, "Proposal was rejected");
```

Comment:
would it be better to check that `price==1e18`? is there a 0.5e18 case or something else in the UMIP that is not approved but also not rejected? put differently it might be better to check for the exact state you're looking for.

Reply:
Good point, we can check that the proposal was approved rather than the other way around. The only approved return value is 1e18.